src/gridcoin/beacon.cpp

```diff
@@ -1463,9 +1582,11 @@ BeaconRegistry::HistoricalBeaconMap::iterator BeaconRegistry::BeaconDB::advance(
     return ++iter;
 }
 
-Beacon& BeaconRegistry::BeaconDB::operator[](const uint256& hash)
+Beacon_ptr& BeaconRegistry::BeaconDB::operator[](const uint256& hash)
 {
-    return m_historical[hash];
+    auto iter = find(hash);
+
+    return iter->second;
```

Comment:
`BeaconDB::find()` can return an iterator for `m_historical.end()` which will cause this to dereference an invalid iterator.

Reply:
I don't understand your point here. Find **should** return .end() if there is no element found in the map or leveldb. I don't understand why you think the iterator is invalid. The iterator was set to auto iter = m_historical.find(hash); at the very beginning. There is a test as to whether it is at .end() from the find (still valid) and then if it is, a load into a beacon OBJECT is attempted from leveldb. If the load is SUCCESSFUL, then a shared pointer to the object is inserted into the map and the iterator updated to reference the iterator from the insert.

Conversely, if the Load() from leveldb into the beacon object is unsuccessful, the iterator remains untouched from what it was originally, m_historical.end().

I really don't see the issue you are driving at here.

Reply:
The expression `iter->second` in the return statement invokes undefined behavior if `BeaconDB::find()` returns `m_historical.end()`. It will likely segfault. The method can return a null `shared_ptr` to satisfy the interface in this case (the return type must change from a reference to a value). However, some may argue that that doing so isn't idiomatic behavior for `operator[]` which customarily initializes the missing element if it doesn't exist and returns a reference to the inserted value. I don't think doing so is useful here, but, in either case, we pass along the responsibility for guarding against a null pointer to the calling code. It may be more clear to use `find()` directly to avoid the ambiguity.

Reply:
Ah... your issue is with the operator[] function itself. I think you are right. Let me go check. We probably don't even need it. I could code it to do the usual thing to insert the missing element, but I didn't intend on providing the "write" side of the operator functionality anyway. It is unnecessary complexity that won't even be used. I think I should simply remove it and exclusively use the find function I implemented instead.

Reply:
Simpler is better. I removed the operator[] function, since it would be complex to implement and that functionality is not even really used. I replaced the use of [] with find()->second.